ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestResourceRecordSetApi.java

```diff
@@ -0,0 +1,205 @@
+import static denominator.common.Preconditions.checkNotNull;
+import static denominator.common.Util.nextOrNull;
+import static denominator.common.Util.toMap;
+
+final class UltraDNSRestResourceRecordSetApi implements denominator.ResourceRecordSetApi {
+
+  private static final int DEFAULT_TTL = 300;
+  private final UltraDNSRest api;
+  private final String zoneName;
+  private final UltraDNSRestRoundRobinPoolApi roundRobinPoolApi;
+
+  UltraDNSRestResourceRecordSetApi(UltraDNSRest api, String zoneName,
+                                   UltraDNSRestRoundRobinPoolApi roundRobinPoolApi) {
+    this.api = api;
+    this.zoneName = zoneName;
+    this.roundRobinPoolApi = roundRobinPoolApi;
+  }
+
+  @Override
+  public Iterator<ResourceRecordSet<?>> iterator() {
+    Iterator<Record> orderedRecords = api
+            .getResourceRecordsOfZone(zoneName)
+            .buildRecords()
+            .iterator();
+    return new GroupByRecordNameAndTypeCustomIterator(orderedRecords);
+  }
+
+  @Override
+  public Iterator<ResourceRecordSet<?>> iterateByName(String name) {
+    checkNotNull(name, "name");
+    final int ANY_RR_TYPE = 255;
+    Iterator<Record> ordered = api
+            .getResourceRecordsOfDNameByType(zoneName, name, ANY_RR_TYPE)
+            .buildRecords()
+            .iterator();
+    return new GroupByRecordNameAndTypeCustomIterator(ordered);
+  }
+
+  @Override
+  public ResourceRecordSet<?> getByNameAndType(String name, String type) {
+    checkNotNull(name, "name");
+    checkNotNull(type, "type");
+    Iterator<Record> orderedRecords = recordsByNameAndType(name, type)
+            .iterator();
+    return nextOrNull(new GroupByRecordNameAndTypeCustomIterator(
+            orderedRecords));
+  }
+
+  private List<Record> recordsByNameAndType(String name, String type) {
+    checkNotNull(name, "name");
+    checkNotNull(type, "type");
+    int typeValue = checkNotNull(lookup(type), "typeValue for %s", type);
+    List<Record> records = null;
+    try {
+      records = api
+              .getResourceRecordsOfDNameByType(zoneName, name, typeValue)
+              .buildRecords();
+    } catch (UltraDNSRestException e) {
+      if (e.code() == UltraDNSRestException.DATA_NOT_FOUND ||
+          e.code() == UltraDNSRestException.RESOURCE_RECORD_POOL_NOT_FOUND) {
+        records = new ArrayList<Record>();
+      } else {
+        throw e;
+      }
+    }
+    return records;
+  }
+
+  @Override
+  public void put(ResourceRecordSet<?> rrset) {
+    checkNotNull(rrset, "rrset was null");
+    checkArgument(!rrset.records().isEmpty(), "rrset was empty %s", rrset);
+    int ttlToApply = rrset.ttl() != null ? rrset.ttl() : DEFAULT_TTL;
+
+    List<Record> toUpdate = recordsByNameAndType(rrset.name(), rrset.type());
+    List<Map<String, Object>> toCreate = new ArrayList<Map<String, Object>>(rrset.records());
+
+    for (Iterator<Record> shouldUpdate = toUpdate.iterator(); shouldUpdate.hasNext();) {
```

Comment:
This is existing logic, not something that I wrote from scratch. To stay as close to the original behaviour as possible, we reused (from SOAP) as much as possible. You suggested the logic you mentioned in your comment and email to Arghya when you tried to clarify his doubt. By then, I had completed this as I didn't face any problems with this.

We executed a few integration test cases on the existing SOAP and REST module code and verified the results. We got the expected results. Full fledged integration testing is still pending though. It will be performed once the code review-modifications-merge cycle is done.